src/main/scala/mesosphere/marathon/core/externalvolume/impl/providers/DVDIProvider.scala

```diff
@@ -0,0 +1,265 @@
+    */
+  private[this] def accepts(volume: ExternalVolume): Boolean = {
+    volume.external.provider == name
+  }
+
+  private[this] def optionalOption(opts: Map[String, String], v: Validator[String]): Validator[String] =
+    new Validator[String] {
+      override def apply(optionName: String): Result = {
+        validate(opts.get(optionName))(optional(v)) match {
+          case Success     => Success
+          // we do this to get sane error messages
+          case Failure(vs) => Failure(vs.map(_.withDescription(optionName)))
+        }
+      }
+    }
+
+  private[this] lazy val validRexRayOptions: Validator[Map[String, String]] = validator[Map[String, String]] { opts =>
+    // there's probaby a better way to do this; this worked and didn't require me to duplicate option keys
+    "dvdi/volumetype" as "" is valid(optionalOption(opts, validLabel))
+    "dvdi/newfstype" as "" is valid(optionalOption(opts, validLabel))
+    "dvdi/iops" as "" is valid(optionalOption(opts, validNaturalNumber))
+    "dvdi/overwritefs" as "" is valid(optionalOption(opts, validBoolean))
+  }
+
+  private[this] def nameOf(vol: ExternalVolumeInfo): Option[String] = {
+    Some(vol.provider + "::" + vol.name)
+  }
+
+  private[this] case object haveOnlyOneReplica extends Validator[AppDefinition] {
+    override def apply(app: AppDefinition): Result =
+      if (app.externalVolumes.nonEmpty && app.instances > 1)
+        Failure(Set(RuleViolation(app.id,
+          s"Number of instances is limited to 1 when declaring DVDI volumes in app ${app.id}", None
+        )))
+      else Success
+  }
+
+  private[this] case object haveUniqueExternalVolumeNames extends Validator[AppDefinition] {
+    override def apply(app: AppDefinition): Result = {
+      val conflicts = volumeNameCounts(app).filter(_._2 > 1).keys
+      if (conflicts.isEmpty) Success
+      else Failure(conflicts.toSet[String].map { e =>
+        RuleViolation(app.id, s"Requested DVDI volume '$e' is declared more than once within app ${app.id}", None)
+      })
+    }
+  }
+
+  // for now this matches the validation for resident tasks, but probably won't be as
+  // restrictive in the future.
+  private[this] lazy val validUpgradeStrategy = validator[UpgradeStrategy] { strategy =>
+    strategy.minimumHealthCapacity should be <= 0.5
+    strategy.maximumOverCapacity should equalTo(0.0)
+  }
+
+  /** @return a count of volume references-by-name within an app spec */
+  private[this] def volumeNameCounts(app: AppDefinition): Map[String, Int] =
+    app.externalVolumes.flatMap{ v => nameOf(v.external) }.groupBy(identity).mapValues(_.size)
+
+  private[this] lazy val validMesosVolume = validator[ExternalVolume] { v =>
+    v.mode is equalTo(Mode.RW)
+  }
+
+  private[this] lazy val validMesosContainer = validator[Container] { ct =>
+    ct.volumes is every(valid(ifDVDIVolume(validMesosVolume)))
+  }
+
+  /*
+  // TODO(jdef) Unclear where this requirement came from. Seems super-useful to mix and match
```

Comment:
That was there before my refactoring. We actually have no idea how this will be with more than one provider. Docker only supports one driver at a time at least.